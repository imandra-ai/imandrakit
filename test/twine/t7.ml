type 'a foo = {
  x: 'a option;
  y: (int * 'a) list;
}
[@@deriving twine, show, eq]

type ('a, 'b) bar = {
  f1: 'a foo list;
  f2: 'b foo list;
}
[@@deriving twine, show, eq]

type bar2 = (int, bool) bar [@@deriving show, twine, eq]

let b : bar2 =
  {
    f1 = [ { x = Some 1; y = [ 0, 0; 1, 1; 2, 2 ] } ];
    f2 = [ { x = None; y = [] }; { x = Some true; y = [ 0, true; 1, false ] } ];
  }
;;

Format.printf "b: %a@." pp_bar2 b

let s = Imandrakit_twine.Encode.to_string bar2_to_twine b;;

Format.printf "len=%d@." (String.length s)

let b' = Imandrakit_twine.Decode.decode_string bar2_of_twine s;;

Format.printf "b': %a@." pp_bar2 b';;
assert (equal_bar2 b b')
