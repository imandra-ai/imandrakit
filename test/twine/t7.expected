b: { T7.f1 = [{ T7.x = (Some 1); y = [(0, 0); (1, 1); (2, 2)] }];
     f2 =
     [{ T7.x = None; y = [] };
       { T7.x = (Some true); y = [(0, true); (1, false)] }]
     }
len=47
len as marshal=56
dump:
[0x0]: []
[0x1]: Array(2) [null,@0x0]
[0x4]: Array(2) [0,true]
[0x7]: Array(2) [1,false]
[0xa]: Array(2) [@0x4,@0x7]
[0xd]: Array(1) [true]
[0xf]: Array(2) [@0xd,@0xa]
[0x12]: Array(2) [@0x1,@0xf]
[0x16]: Array(2) [0,0]
[0x19]: Array(2) [1,1]
[0x1c]: Array(2) [2,2]
[0x1f]: Array(3) [@0x16,@0x19,@0x1c]
[0x23]: Array(1) [1]
[0x25]: Array(2) [@0x23,@0x1f]
[0x28]: Array(1) [@0x25]
[0x2a]: Array(2) [@0x28,@0x12]

b': { T7.f1 = [{ T7.x = (Some 1); y = [(0, 0); (1, 1); (2, 2)] }];
      f2 =
      [{ T7.x = None; y = [] };
        { T7.x = (Some true); y = [(0, true); (1, false)] }]
      }
